### PR TITLE
Add option to wait for db to become available

### PR DESCRIPTION
**Pull request by jdavcs**
Configurable option to wait for the database to become available instead
of failing immediately.

_configure_models() seems to be the earliest/reasonable place for it.

### CODE REVIEW

File: config/galaxy.yml.sample [hunk 1]
```diff
@@ -163,6 +163,15 @@ galaxy:
   # not recommended for production use.
   #database_auto_migrate: false
 
+  # Wait for database to become available instead of failing immediately.
+  #database_wait: false
+
+  # Number of attempts before failing.
+  #database_wait_attempts: 60
+
+  # Sleep period between attepmts (seconds; can be a floating point number).
```
Comment (nsoranzo) [inline config/galaxy.yml.sample]:
```suggestion
  # Sleep period between attempts (seconds; can be a floating point number).
```

**Comment by nsoranzo** [comment]
`config/galaxy.yml.sample` should not be modified directly, you should update `lib/galaxy/webapps/galaxy/config_schema.yml` and then execute `make config-rebuild` .

File: lib/galaxy/config.py [hunk 2]
```diff
@@ -1196,3 +1206,15 @@ def _configure_models(self, check_migrate_databases=False, check_migrate_tools=F
     def _configure_signal_handlers(self, handlers):
         for sig, handler in handlers.items():
             signal.signal(sig, handler)
+
+    def _wait_for_database(self, url):
+        from time import sleep
```
Comment (nsoranzo) [inline lib/galaxy/config.py]: I'd move the standard library import to the top of the file (line 21).

Comment (jdavcs) [reply lib/galaxy/config.py]: I would too, no question about that! Was trying to be consistent with the rest of the file (and tread softly). So yes, will do.

File: config/galaxy.yml.sample (hunk 1 above)
Comment (jdavcs) [reply config/galaxy.yml.sample]: and thanks! :)

File: lib/galaxy/config.py (hunk 2 above)
Comment (jdavcs) [reply lib/galaxy/config.py]: We can do without it: time module is already imported, so just time.sleep(pause).
I'm leaving the sqlalchemy_utils import in the method.

Comment (nsoranzo) [reply lib/galaxy/config.py]: > We can do without it: time module is already imported, so just time.sleep(pause).

Surely better!

Comment (jdavcs) [reply lib/galaxy/config.py]: Done. Thank you for reviewing!

Comment (jdavcs) [reply lib/galaxy/config.py]: Correction: not done. I've just noticed your comment (that I missed) on generating config/galaxy.yml.sample. Working on it...

**Review by nsoranzo — APPROVED** [review]

**Comment by nsoranzo** [comment]
@ic4f Still WIP?

File: lib/galaxy/config.py [hunk 3]
```diff
@@ -1196,3 +1206,14 @@ def _configure_models(self, check_migrate_databases=False, check_migrate_tools=F
...
         for sig, handler in handlers.items():
             signal.signal(sig, handler)
+
+    def _wait_for_database(self, url):
+        from sqlalchemy_utils import database_exists
+        attempts = self.config.database_wait_attempts
+        pause = self.config.database_wait_sleep
+        for i in range(1, attempts):
+            try:
+                database_exists(url)
```
Comment (nuwang) [inline lib/galaxy/config.py:1216]: Does this need a break here so the loop doesn't continue if the database exists?

Comment (jdavcs) [reply lib/galaxy/config.py:1216]: Of course! (yikes... missed this one!) Thank you! Pushing...
The rest - done, no longer WIP (config schema updated, make config rebuild done).

**Review by nsoranzo — APPROVED** [review]

**Comment by dannon** [comment]
Selenium error is unrelated.  Thanks @ic4f!

**Pull request merged**